Make L2ARC tests more robust

gamanakis opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Buildbot%20Options.html
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Fixes #10981 

### Description
<!--- Describe your changes in detail -->
Instead of relying on arbitrary timers after pool export/import or cache
device off/online, rely on arcstats. This makes the tests more robust.
Also cleanup some functions related to persistent L2ARC.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

gamanakis commented:
Local testing revealed that the persist_l2arc_007 was failing with a rate 2.5% (~1/40 runs). With this PR this is 0%, locally tested in multiple runs (500 runs). 

@behlendorf Could you restart the failing buildbots? Also is there a way to see how often it failed in buildbots building the master branch?

behlendorf commented:
@behlendorf Could you restart the failing buildbots?

Done. 

> Also is there a way to see how often it failed in buildbots building the master branch?

In fact there is.  We track the ZTS failures in the master branch [here](http://build.zfsonlinux.org/known-issues.html).  It looks like 5 / 316 runs on master, or about 1.55% (not counting PRs).

behlendorf commented:
Looking at the full list I see we actually did get a couple other l2arc ZTS failures, but they were a bit more rare.

0.65% - 2 / 304 - [11898](http://build.zfsonlinux.org/builders/Amazon%202%20x86_64%20Release%20%28TEST%29/builds/11898) [1205](http://build.zfsonlinux.org/builders/Fedora%2032%20x86_64%20%28TEST%29/builds/1205) - tests/functional/l2arc/l2arc_mfuonly_pos
0.32% - 1 / 305 - [5006](http://build.zfsonlinux.org/builders/Debian%2010%20x86_64%20%28TEST%29/builds/5006) - tests/functional/l2arc/persist_l2arc_008_pos
0.32% - 1 / 305 - [1205](http://build.zfsonlinux.org/builders/Fedora%2032%20x86_64%20%28TEST%29/builds/1205) - tests/functional/l2arc/l2arc_arcstats_pos

gamanakis commented:
> Looking at the full list I see we actually did get a couple other l2arc ZTS failures, but they were a bit more rare.
> 
> 0.65% - 2 / 304 - [11898](http://build.zfsonlinux.org/builders/Amazon%202%20x86_64%20Release%20%28TEST%29/builds/11898) [1205](http://build.zfsonlinux.org/builders/Fedora%2032%20x86_64%20%28TEST%29/builds/1205) - tests/functional/l2arc/l2arc_mfuonly_pos
> 0.32% - 1 / 305 - [5006](http://build.zfsonlinux.org/builders/Debian%2010%20x86_64%20%28TEST%29/builds/5006) - tests/functional/l2arc/persist_l2arc_008_pos
> 0.32% - 1 / 305 - [1205](http://build.zfsonlinux.org/builders/Fedora%2032%20x86_64%20%28TEST%29/builds/1205) - tests/functional/l2arc/l2arc_arcstats_pos

I saw these too, took a look and in 3/4 of those failures it seems unable to grab the arcstat, I don't know why this would happen. It doesn't seem to be a problem with the test per se. The remaining failure should be addressed by this PR.

behlendorf commented on module/zfs/arc.c:
```diff
@@ -9975,7 +9970,7 @@ l2arc_dev_hdr_read(l2arc_dev_t *dev)
 	err = zio_wait(zio_read_phys(NULL, dev->l2ad_vdev,
 	    VDEV_LABEL_START_SIZE, l2dhdr_asize, abd,
-	    ZIO_CHECKSUM_LABEL, NULL, NULL, ZIO_PRIORITY_ASYNC_READ,
+	    ZIO_CHECKSUM_LABEL, NULL, NULL, ZIO_PRIORITY_SYNC_READ,
```
This makes sense, was the use of async here merely causing a reporting issue?

gamanakis replied:
No it wasn't causing any issue. This is a synchronous read (zio_wait) and I thought it would be more correct to have the corresponding flag, or am I wrong?

behlendorf replied:
No you're right.  I just thought you might have noticed it was being reported incorrectly in something like thew `zpool status -r` output which breaks down the IOs in part by priority.  Sounds like you just noticed it by inspection which is great.

behlendorf commented on tests/zfs-tests/tests/functional/l2arc/l2arc_arcstats_pos.ksh:
```diff
@@ -70,19 +70,59 @@ log_must fio $FIO_SCRIPTS/mkfiles.fio
+	sleep 2
+	typeset l2size2=$(get_arcstat l2_size)
+	do_once=false
+done
```
nit: since all of these blocks the same moving this to a helper function might be nice.  Many of the test groups include a shared `.kshlib` file which is sourced to provide some useful functions.

gamanakis replied:
Yes, will do!

behlendorf reviewed (commented):
Looks good.  Thanks for running down these issues.

gamanakis commented:
7de9372: Increase the test threshold in l2arc_mfuonly_pos test and explain why (it's also in the man page of l2arc_mfuonly). More importantly implement a function in `libtest.shlib` that l2arc tests can leverage instead of relying on timers.

adamdmoss commented on tests/zfs-tests/tests/functional/l2arc/persist_l2arc_006_pos.ksh:
```diff
@@ -73,24 +73,24 @@ log_must fio $FIO_SCRIPTS/random_reads.fio
+	$l2_rebuild_log_blk_start ))
 log_must test $l2_dh_log_blk -gt 0
 
+log must zpoool offline $TESTPOOL $VDEV_CACHE
```
'zpoool'

gamanakis replied:
I though I had fixed that! good catch!

adamdmoss commented on tests/zfs-tests/include/libtest.shlib:
```diff
@@ -4173,6 +4173,24 @@ function get_arcstat # stat
+	typeset stat=$1
+	typeset do_once=true
+
+	while $do_once || [ $stat1 -ne $stat2 ] || [ $stat2 -eq 0 ]; do
```
For stats that can go up or down, technically this will find either a plateau or a, err, valley.   Quiescence.  Is that the intent?

gamanakis replied:
Essentially inactivity, does not matter if plateau or valley. Quiescence. No activity on L2ARC devices judged by l2_size or complete removal of the L2ARC device as seen by l2_feeds. I will rename it to arcstat_quiescence.

adamdmoss commented on tests/zfs-tests/tests/functional/l2arc/l2arc_mfuonly_pos.ksh:
```diff
@@ -75,7 +75,13 @@ log_must fio $FIO_SCRIPTS/random_reads.fio
+# If accessed again they are counted as MRU and the l2arc_mru_asize arcstat
+# will not be 0. That is the reason with we set the following test to less
+# than 1kB and not exactly 0.
+log_must test $(get_arcstat l2_mru_asize) -le 1024
```
I appreciate the explanation but I'm not sure it really explains where the magic 1kB comes from - does the test expect <1kB recordsize or something?

adamdmoss replied:
"it's also in the man page of l2arc_mfuonly" - sad to say I'm not sure which man page this is within :(

gamanakis replied:
the man page of l2arc_mfuonly of course, in zfs-module-parameters. I will clarify.

gamanakis replied:
The magic value of 1kB is based on multiple runs. When it fails, it fails with 512 in my testing, so I doubled that value.

gamanakis replied:
Perhaps I can fix this by setting zfs_prefetch_disable=1 in the test. Let me test.

gamanakis replied:
In a local run of 1000 tests, setting zfs_prefetch_disable and restoring testing for 0 (as opposed to 1kB) produced no failures. Adopted this approach and updated the comments.

adamdmoss commented on tests/zfs-tests/include/libtest.shlib:
```diff
@@ -4173,6 +4173,24 @@ function get_arcstat # stat
+		do_once=false
+	done
+
+	echo $stat2
```
Did you mean to leave this echo of raw $stat2 in?

adamdmoss replied:
Ah, I see, some users of arcstat_plateau use it as a value.   (Will this leave junk on stdout when they don't?)

gamanakis replied:
It seems it does, will fix.

adamdmoss commented on tests/zfs-tests/include/libtest.shlib:
```diff
@@ -4173,6 +4173,24 @@ function get_arcstat # stat
+	done
+
+	echo $stat2
+}
```
Since there's no upper bound on how long this helper waits for a settle, should it echo something like 'waiting for $stat to plateau...' when it starts, to make it easy to see what happened when a test times out?

adamdmoss replied:
I guess that's in conflict with being able to use the interpolated output of the helper.  I'd like to see a wrapper for standalone calls as a separate thing from one used for interpolation, but it's not a biggie.

gamanakis replied:
I can easily fix this by inserting echoes right before the call to the helper in the tests.

gamanakis replied:
Or I can implement another helper like arcstat_quiescence_echo.

gamanakis commented:
e971468: addressed comments from @adamdmoss , thank you! Renamed arcstat_plateau to arcstat_quiescence, implement two variants of that function (one for interpolation, the other for standalone calls), clarify the comment in l2arc_mfuonly_pos.

gamanakis commented:
@freqlabs what is the equivalent of zfs_prefetch_disable in FreeBSD?

freqlabs commented:
> @freqlabs what is the equivalent of zfs_prefetch_disable in FreeBSD?

`vfs.zfs.prefetch.disable`, so in tunables.cfg it will be `prefetch.disable`

gamanakis commented:
8be9ba0: fixed FreeBSD tunable (zfs_prefetch_disable, prefetch.disable).

adamdmoss reviewed (approved):
nice!

gamanakis commented:
The persist_l2arc_008_pos failed in the Debian 10 buildbot. Let me take a look.

gamanakis commented:
be6f925: In all tests additionally wait for l2_size to become quiescent after importing the pool or before off-lining a cache device.

gamanakis commented:
d3f5dac: Accidentally broke the persist_l2arc_001 test. Fixed.

behlendorf commented:
@gamanakis I resubmitted the builds and everything looks good.  If you don't have any more changes I'd like to move forward with this.

gamanakis commented:
@behlendorf I think it is good now. You can go ahead. Thank you!
